codegen/src/api/mod.rs

```diff
@@ -378,84 +342,91 @@ where
         .variants
         .iter()
         .map(|var| {
+            let mut type_params = TypeParameters::from_scale_info(&[]);
+            let composite_ir_kind =
+                type_gen.create_composite_ir_kind(&var.fields, &mut type_params)?;
             let struct_name = variant_to_struct_name(&var.name);
-
-            let fields = CompositeDefFields::from_scale_info_fields(
-                struct_name.as_ref(),
-                &var.fields,
-                &[],
-                type_gen,
-            )?;
-
-            let alias_module_name = format_ident!("{}", var.name.to_snake_case());
-
-            let docs = should_gen_docs.then_some(&*var.docs).unwrap_or_default();
-            let struct_def = CompositeDef::struct_def(
-                &ty,
-                struct_name.as_ref(),
-                Default::default(),
-                fields.clone(),
-                Some(parse_quote!(pub)),
-                type_gen,
-                docs,
-                crate_path,
-                Some(alias_module_name.clone()),
-            )?;
-
-            let type_aliases = TypeAliases::new(fields, types_mod_ident.clone(), alias_module_name);
-
-            Ok((var.name.to_string(), struct_def, type_aliases))
+            let mut composite = CompositeIR::new(
+                syn::parse_str(&struct_name).expect("enum variant name is valid ident"),
```

Comment:
This should be infallible

Reply:
Yes, unless the metadata is somehow corrupted. Should I formulate the `expect` message differently here?

Reply:
I prefer what we use in parity with a explanation/"semi-formal" proof why it can't panic such `"enum variant is a valid; qed"` that is very clear to me :)